sklearn/linear_model/_glm/_newton_solver.py

```diff
@@ -516,3 +532,460 @@ def inner_solve(self, X, y, sample_weight):
+
+    with
+
+        G = X.T @ g + l2_reg_strength * coef
+        H = X.T @ diag(h) @ X + l2_reg_strength * P
+        g = loss.gradient = pointwise gradient
+        h = loss.hessian = pointwise hessian
+        P = penalty matrix in 1/2 w @ P @ w,
+            for a pure L2 penalty without intercept it equals the identity matrix.
+
+    is cast as a least squares problem
+
+        min ||A @ coef_newton - b||_2^2
+
+    with
+
+        A = [                    sqrt(h) * X]
+            [sqrt(l2_reg_strength) * sqrt(P)]
+        b = [                                - g / sqrt(h)]
+            [- sqrt(l2_reg_strength) * sqrt(P) @ self.coef]
+
+    Notes:
+    - A is a square root of H: H = A'A
+    - A and b form G: G = -A'b
+    - The normal equation of this least squares problem, A'A coef_newton = A'b, is again
+      H @ coef_newton = -G.
+
+    Note that this solver can naturally deal with sparse X.
+
+    References
+    ----------
+    .. [1] :arxiv:`Fong & Saunders "LSMR: An iterative algorithm for sparse
+           least-squares problems" <1006.0758>`
+           See also
+           https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.lsmr.html
+
+    """  # noqa: E501
+
+    def setup(self, X, y, sample_weight):
+        """Setup.
+
+        Additionally to super().setup(), also sets:
+            - self.g = pointwise gradient
+            - self.h = pointwise hessian
+            - self.gradient
+            - self.sqrt_P = sqrt(l2_reg_strength) * sqrt(P)
+            - self.A_norm
+        """
+        super().setup(X=X, y=y, sample_weight=sample_weight)
+        (self.g, self.h,) = self.linear_loss.base_loss.init_gradient_and_hessian(
+            n_samples=X.shape[0], dtype=X.dtype
+        )
+        n_features = X.shape[1]
+        # For a general (symmetric) penalty matrix P, we can use any square root of it,
+        # e.g. the Cholesky decomposition. For the simple L2-penalty, we can use the
+        # identity matrix and handle the intercept separately.
+        # We use a 1-d array of shape (n_features,) instead of a 2-d diagonal array to
+        # save memory. We also simply omit the zero at the end for the intercept.
+        # For the multiclass case, we use the same 1-d array of shape (n_features,) and
+        # do not store n_classes copies of it.
+        self.sqrt_P = np.full(
+            shape=n_features, fill_value=np.sqrt(self.l2_reg_strength), dtype=X.dtype
+        )
+
+        # Initially, set Frobenius norm of A as if h = 1
+        # => ||A||^2 = ||X||^2 + sqrt(l2_reg_strength) ||sqrt(P)||^2
+        # if scipy.sparse.issparse(X):
+        #     A_norm = scipy.sparse.linalg.norm(X) ** 2
+        # else:
+        #     A_norm = scipy.linalg.norm(X) ** 2
+        # A_norm += scipy.linalg.norm(self.sqrt_P) ** 2
+        # self.A_norm = np.sqrt(A_norm)
+        # As this all is a bit costly, we make life easier and start with 1:
+        self.A_norm = 1
+        self.r_norm = 1
```

Comment:
Why is this valid?

Reply:
Because it works😏 Without penalty and with $h=1$, one has $||A|| = ||X||$. We likely have $||X|| > 1$, so `atol ~ 1/||A||` is a looser criterion with an underestimated $||A||$ which is a good choice for the first iteration.

Reply:
Done in https://github.com/scikit-learn/scikit-learn/pull/25462/commits/59c13221b94c3e058d087446aa949deb3f049f83.